src/test/groovy/net/schmizz/sshj/transport/verification/KeyWithCertificateUnitSpec.groovy

```diff
@@ -0,0 +1,221 @@
+                .pubKey
+
+        and:
+        Buffer certRaw = new Buffer.PlainBuffer();
+        certRaw.putPublicKey(hostCertificate);
+
+        when:
+        String error = KeyType.CertUtils.verifyHostCertificate(certRaw.getCompactData(), hostCertificate, "127.0.0.1")
+
+        then:
+        error != null && error.startsWith("Certificate is valid after ")
+    }
+
+    def "verifying an valid certificate with no principal"() {
+        given:
+        def hostCertificate = (Certificate<PublicKey>) OpenSSHKeyFileUtil
+                .initPubKey(new FileReader(
+                        new File("src/itest/resources/keyfiles/certificates/" +
+                                 "id_ed25519_384_rfc4716_signed_by_rsa_host_no_principal-cert.pub")))
+                .pubKey
+
+        and:
+        Buffer certRaw = new Buffer.PlainBuffer();
+        certRaw.putPublicKey(hostCertificate);
+
+        when:
+        String error1 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "good.example.com")
+        String error2 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "127.0.0.1")
+        String error3 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "good.example.bad.com")
+
+        then:
+        error1 == null
+        error2 == null
+        error3 == null
+    }
+
+    def "verifying an valid certificate with wildcard principal"() {
+        given:
+        def hostCertificate = (Certificate<PublicKey>) OpenSSHKeyFileUtil
+                .initPubKey(new FileReader(
+                        new File("src/itest/resources/keyfiles/certificates/" +
+                                 "id_ed25519_384_rfc4716_signed_by_rsa_host_principal_wildcard_example_com-cert.pub")))
+                .pubKey
+
+        and:
+        Buffer certRaw = new Buffer.PlainBuffer();
+        certRaw.putPublicKey(hostCertificate);
+
+        when:
+        String error1 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "good.example.com")
+        String error2 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "127.0.0.1")
+        String error3 = KeyType.CertUtils.verifyHostCertificate(
+                certRaw.getCompactData(), hostCertificate, "good.example.bad.com")
+
+        then:
+        error1 == null
+        error2 != null && error2.contains("doesn't match any of the principals")
+        error3 != null && error3.contains("doesn't match any of the principals")
+    }
+
+    def "KeyType CertUtils checkPrincipals"() {
+        // Based on regress/unittests/match/test.c of portable OpenSSH, commit 068dc7ef783d135e91ff954e754d2ed432e
+        expect:
+        KeyType.CertUtils.matchPattern("", "")
```

Comment:
Apparently, these test cases were taken from [the code under BSD license](https://github.com/openssh/openssh-portable/blob/master/LICENCE). Although it must be fine to reuse BSD code inside Apache 2.0 project, I'm not sure if some license text should be added there or not.

Reply:
Not sure either. They're not literal copies I think, only the strings for the cases.

Reply:
So, I replaced `ab` to `xy`, I hope it's enough.